src/refactor_nrepl/artifacts.clj

```diff
@@ -128,19 +138,88 @@
     (update-artifact-cache!))
   (->> @artifacts keys list*))
 
+(defn- artifact-versions* [artifact-id]
+  (->> (or (get @artifacts artifact-id)
+           (seq (get-mvn-versions! artifact-id))
+           (get-clojars-versions! artifact-id))
+       distinct
+       versions/version-sort
+       reverse
+       list*))
+
 (defn artifact-versions
   "Returns a sorted list of artifact version strings. The list can either come
   from the artifacts cache, the maven search api or the clojars search api in
   that order."
   [{:keys [artifact]}]
-  (->> (or (get @artifacts artifact)
-           (seq (get-mvn-versions! artifact))
-           (get-clojars-versions! artifact))
-       distinct
-       versions/version-sort
-       reverse
-       list*))
+  (artifact-versions* artifact))
+
+(defn- jar-at-the-top-of-dependency-hierarchy [new-deps]
+  ;; We only need to consider the dep at the top of the hierarchy because when we
+  ;; require those namespaces the rest of  the transitive deps will get pulled in
+  ;; too.
+  (letfn [(->jar [^File f]
+            (JarFile. f))]
+    (let [top-level-dep  (-> new-deps keys first)]
+      (-> top-level-dep meta :file ->jar))))
+
+(defn- make-resolve-missing-aware-of-new-deps!
+  "Once the deps are available on cp we still have to load them and
+  reset slamhound's cache to make resolve-missing work."
+  [^JarFile jar]
+  (doseq [new-namespace (find/find-namespaces-in-jarfile jar)]
+    (try
+      (require new-namespace)
+      (catch Exception _
+        ;; I've seen this happen after adding core.async as a dependency.
+        ;; It also happens if you try to require namespaces that no longer work,
+        ;; like compojure.handler.
+        ;; A failure here isn't a big deal, it only means that resolve-missing
+        ;; isn't going to work until the namespace has been loaded manually.
+        )))
+  (slamhound/reset)
+  (slamhound-regrow/clear-cache!))
+
+(defn- parse-coordinates [coordinates-str]
+  (let [coords (try (->> coordinates-str edn/read-string (take 2) vec)
+                    (catch Exception _))]
+    (if (and (= (count coords) 2)
```

Comment:
will the `2` constraint work in face of :exclusions, :classifiers etc?

Reply:
I don't think it makes sense to handle act on `:exclusions`.  My thinking is that you'd never need to hotload a dep that's excluded because

1. It's excluded because it's never needed
2. It gets (hot) loaded through some other artifact and was excluded to prevent ambiguity 

Finally, a request to simply load `[foo/bar]` isn't something that we should be willing to act on because it would be ambiguous.  Do they want any old version?  The latest stable?  The latest snapshot?

> :classifiers 

Didn't know about this and have never encountered it in the wild. def something we could support.

Reply:
Oh I think I wasn't specific enough, sorry for that

These are the sorts of inputs I had in mind:

* `[com.foo/bar "1.2.3" :exclusions [xyz]]`
* `[com.foo/bar "1.2.3" :classifier "test"]`

Both are things that can be passed straight to pomegranate IIRC. Not only we prevent a precondition from failing too easily, but we also make pomegranate resolve the right artifact, with the right exclusions (which otherwise could foil a classpath) etc